lib/generator.ml

```diff
@@ -136,29 +136,46 @@ let get_version =
   {|let get_version s = 
   get_version_from_json (Yojson.Safe.from_string s)|}
 
-let get_version_from_lexbuf =
-  {|let get_version_from_lexbuf p lb = 
-  get_version_from_json (Yojson.Safe.from_lexbuf p lb)|}
+let get_version_from_file =
+  {|let get_version_from_file fname = 
+  get_version_from_json (Yojson.Safe.from_file ~fname fname)|}
 
-let make_decode_main ~impl_kind ~prefix ~fn_name ~arg ~fn_sig ~match_version
+let make_decode_main
+    ~impl_kind ~prefix ~decode ~fn_name ~arg ~fn_sig ~match_version
   = function
   | [] ->
     assert false
   | latest_version :: tail ->
     let version_matches =
       List.map
         ~f:(fun version ->
+          let module_name =
+            Upgrader.name_impl_module impl_kind prefix version
+          in
+          let decode =
+            match decode with
+            | Some decode ->
+              decode module_name
+            | None ->
+              [%string "$(module_name).$fn_name $arg"]
```

Comment:
we can keep this a bit more dry (repetition below)

Reply:
I think it's a bit better now, but it's still not the code I'm the most proud of haha, it feels a bit ugly overall.